Sources/Moya/Task.swift

```diff
@@ -3,39 +3,48 @@ import Foundation
-
-    /// A requests body set with encoded parameters.
-    case requestParameters(parameters: [String: Any], encoding: ParameterEncoding)
-
-    /// A requests body set with data, combined with url parameters.
-    case requestCompositeData(bodyData: Data, urlParameters: [String: Any])
-
-    /// A requests body set with encoded parameters combined with url parameters.
-    case requestCompositeParameters(bodyParameters: [String: Any], bodyEncoding: ParameterEncoding, urlParameters: [String: Any])
-
-    /// A file upload task.
-    case uploadFile(URL)
-
-    /// A "multipart/form-data" upload task.
-    case uploadMultipart([MultipartFormData])
-
-    /// A "multipart/form-data" upload task  combined with url parameters.
-    case uploadCompositeMultipart([MultipartFormData], urlParameters: [String: Any])
-
-    /// A file download task to a destination.
-    case downloadDestination(DownloadDestination)
-
-    /// A file download task to a destination with extra parameters using the given encoding.
-    case downloadParameters(parameters: [String: Any], encoding: ParameterEncoding, destination: DownloadDestination)
+    /// All different ways to set parameters in an HTTP request's body.
+    public enum BodyParams {
+        /// The given encodable will be set in the request's body without any additional encoding.
+        case raw(Data)
+
+        /// The given encodable will be json encoded in the request's body.
+        case json(Encodable, JSONEncoder = JSONEncoder())
+
+        /// The given encodable will be url encoded in the request's body.
+        case urlEncoded(Encodable, URLEncodedFormEncoder = URLEncodedFormEncoder())
+
+        /// The given encodable will be encoded in the request's body using the provided encoder.
+        ///
+        /// The provided encoder must not be a `URLEncodedFormParameterEncoder` (use `BodyParams.urlEncoded`instead)
+        /// or `JSONParameterEncoder`(use `BodyParams.json` instead). If this is the case, a `MoyaError.encodableMapping` will be raised.
+        case custom(Encodable, ParameterEncoder)
+    }
+
+    /// All different ways to set parameters in an HTTP request's query.
+    public enum QueryParams {
+        /// The given encodable will be url encoded in the request's query.
+        case query(Encodable, URLEncodedFormEncoder = URLEncodedFormEncoder())
+    }
+
+    /// All sources available for use when uploading
+    public enum UploadSource {
+        /// Upload the provided data
+        case rawData(Data)
+
+        /// Upload the file at the given url
+        case file(URL)
+
+        /// Upload some multipart content.
+        case multipart([MultipartFormData])
+    }
+
+    /// A task to request some data
+    case request(bodyParams: BodyParams? = nil, queryParams: QueryParams? = nil)
+
+    /// A task to upload some data
+    case upload(source: UploadSource, bodyParams: BodyParams? = nil, queryParams: QueryParams? = nil)
```

Comment:
should body parameters be possible to add for an upload? if we want to add parameters to the upload we are doing, we would use multipart/form-data (in that case `.multipart` as a source)

Reply:
That is a good question ^^ I don't have the answer as I never had to upload something.

Reply:
I was thinking more about removing the body parameters as `MultipartFormData` already have an option to provide properties alongside data upload.

Reply:
Ok I didn't understand at first ;)
Task.upload no longer has a BodyParams.